### PR TITLE
feat(hubble): push/pull rich logging

### PR DESCRIPTION
**Pull request by numb3r3**

### CODE REVIEW

**Review by mapleeit — APPROVED** [review]

**Review by winstonww — APPROVED** [review]

**Comment by mapleeit** [comment]
Pulling:

Screen Shot 2021-08-06 at 10 37 25 AM

Pushing:

Screen Shot 2021-08-06 at 10 38 45 AM

File: jina/hubble/hubio.py [hunk 1]
```diff
@@ -575,74 +578,110 @@ def pull(self) -> str:
...
+                                continue
+                            status = log['status']
+                            status_id = log.get('id', None)
+                            pg_detail = log.get('progressDetail', None)
+
+                            if pg_detail is None:
+                                console.print(status)
+                                continue
+
+                            if status_id not in tasks:
```
Comment (JoanFM) [inline jina/hubble/hubio.py:609]: this is not safe if status_id is None

File: jina/hubble/hubio.py [hunk 2]
```diff
@@ -575,74 +578,110 @@ def pull(self) -> str:
...
+                                task = progress.add_task(status, total=0)
+                                tasks[status_id] = task
+
+                            if pg_detail:
```
Comment (JoanFM) [inline jina/hubble/hubio.py:613]:
```suggestion
                            if pg_detail is not None:
```

File: jina/hubble/hubio.py [hunk 3]
```diff
@@ -575,74 +578,110 @@ def pull(self) -> str:
...
+                                task = tasks[status_id]
```
Comment (JoanFM) [inline jina/hubble/hubio.py:614]: what if `status_id` is None

File: jina/hubble/hubio.py [hunk 4]
```diff
@@ -575,74 +578,110 @@ def pull(self) -> str:
...
+            elif scheme == 'jinahub':
+                import filelock
+
+                with filelock.FileLock(get_lockfile(), timeout=-1):
+                    try:
+                        pkg_path, pkg_dist_path = resolve_local(executor)
+                        # check serial number to upgrade
+                        sn_file_path = pkg_dist_path / f'PKG-SN-{executor.sn or 0}'
+                        if not sn_file_path.exists() and any(
+                            True for _ in pkg_dist_path.glob('PKG-SN-*')
```
Comment (JoanFM) [inline jina/hubble/hubio.py:640]: I am not sure if `True` is needed here

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jina/hubble/hubio.py [hunk 5]
```diff
@@ -555,16 +560,14 @@ def pull(self) -> str:
...
-            import docker
+        from rich.progress import Progress, DownloadColumn, BarColumn
 
         console = Console()
         cached_zip_file = None
         usage = None
 
-        with console.status(f'Pulling {self.args.uri}...') as st:
-            try:
+        try:
```
[the remark this answers is not in the document]
Comment (numb3r3) [reply jina/hubble/hubio.py:569]: Thanks for your remind. With manually test, the ctx also raises the exception to the outside

**Comment by numb3r3** [comment]
@JoanFM thanks for your comments. I made a new commit to address them. To make it safe to work with `status_id=None` or `pg_detail = None`, I add if condition checking at the beginning.

**Review by JoanFM — APPROVED** [review]

**Pull request merged**